vsphere/internal/virtualdevice/virtual_machine_device_subresource.go

```diff
@@ -547,57 +548,98 @@ func swapSCSIDevice(l object.VirtualDeviceList, device types.BaseVirtualSCSICont
+	sataCtlrs := make([]types.BaseVirtualSATAController, sataCount)
+	ideCtlrs := make([]*types.VirtualIDEController, ideCount)
 	// Don't worry about doing any fancy select stuff here, just go thru the
 	// VirtualDeviceList and populate the controllers.
-	iCount := int32(count)
+	log.Printf("[DEBUG] NormalizeBus: Normalizing first %d controllers on SCSI bus to device type %s", scsiCount, scsiType)
+	log.Printf("[DEBUG] NormalizeBus: Normalizing first %d controllers on SATA bus", sataCount)
+	log.Printf("[DEBUG] NormalizeBus: Normalizing first %d controllers on IDE bus", ideCount)
 	for _, dev := range l {
-		if sc, ok := dev.(types.BaseVirtualSCSIController); ok {
-			if busNumber := sc.GetVirtualSCSIController().BusNumber; busNumber < iCount {
-				ctlrs[busNumber] = sc
+		switch ctlr := dev.(type) {
+		case types.BaseVirtualSCSIController:
+			if busNumber := ctlr.GetVirtualSCSIController().BusNumber; busNumber < int32(scsiCount) {
+				scsiCtlrs[busNumber] = ctlr
+			}
+		case types.BaseVirtualSATAController:
+			if busNumber := ctlr.GetVirtualSATAController().BusNumber; busNumber < int32(sataCount) {
+				sataCtlrs[busNumber] = ctlr
+			}
+		case *types.VirtualIDEController:
+			if busNumber := ctlr.GetVirtualController().BusNumber; busNumber < int32(ideCount) {
+				ideCtlrs[busNumber] = ctlr
 			}
 		}
 	}
-	log.Printf("[DEBUG] NormalizeSCSIBus: Current SCSI bus contents: %s", scsiControllerListString(ctlrs))
+	log.Printf("[DEBUG] NormalizeBus: Current SCSI bus contents: %s", scsiControllerListString(scsiCtlrs))
 	// Now iterate over the controllers
-	for n, ctlr := range ctlrs {
+	for n, ctlr := range scsiCtlrs {
 		if ctlr == nil {
-			log.Printf("[DEBUG] NormalizeSCSIBus: Creating SCSI controller of type %s at bus number %d", ct, n)
-			cspec, err := createSCSIController(&l, ct, st)
+			log.Printf("[DEBUG] NormalizeBus: Creating SCSI controller of type %s at bus number %d", scsiType, n)
+			cspec, err := createSCSIController(&l, scsiType, scsiSharing)
 			if err != nil {
 				return nil, nil, err
 			}
 			spec = append(spec, cspec...)
 			continue
 		}
-		if l.Type(ctlr.(types.BaseVirtualDevice)) == ct {
-			cspec, err := setSCSIBusSharing(&l, ctlr, st)
+		if l.Type(ctlr.(types.BaseVirtualDevice)) == scsiType {
+			cspec, err := setSCSIBusSharing(&l, ctlr, scsiSharing)
 			if err != nil {
 				return nil, nil, err
 			}
 			spec = append(spec, cspec...)
 			continue
 		}
-		cspec, err := swapSCSIDevice(l, ctlr, ct, st)
+		cspec, err := swapSCSIDevice(l, ctlr, scsiType, scsiSharing)
 		if err != nil {
 			return nil, nil, err
 		}
 		spec = append(spec, cspec...)
 		l = applyDeviceChange(l, cspec)
 		continue
 	}
-	log.Printf("[DEBUG] NormalizeSCSIBus: Outgoing device list: %s", DeviceListString(l))
-	log.Printf("[DEBUG] NormalizeSCSIBus: Outgoing device config spec: %s", DeviceChangeString(spec))
+	log.Printf("[DEBUG] NormalizeBus: Current SATA bus contents: %s", sataControllerListString(sataCtlrs))
+	// Now iterate over the controllers
```

Comment:
I think it would improve readability if the comment said which type of controller we are iterating

Reply:
Good call. Got that fixed.